src/cmd/Makefile.am

```diff
@@ -175,6 +175,7 @@ flux_start_LDADD = \
 
 flux_job_LDADD = \
 	$(top_builddir)/src/common/libsubprocess/libsubprocess.la \
+	$(top_builddir)/src/common/libflux/watcher.lo \
```

Comment:
As a quick experiment I tried moving `fbuf.[ch]` and `fbuf_watcher.[ch]` to libflux, but this just moves the problem, making these symbols unresolved when linking with `libsubprocess.la`. I did verify that can be ameliorated by linking with `libflux/libflux.la` (and I think if this is placed after `libflux-core.la` then only missing symbols are linked instead of all `flux_*` symbols, though I didn't compare final executable sizes or anything.)

Not sure if you see that as an improvement or not.

Reply:
Yeah that was unsatisfying :-(

I think the fbuf watcher and sdbus watcher are appropriate to keep private to libsubprocess and sdbus as they are not really designed for wider usage.  Not that moving them necessarily would mean that I guess.

I'll try your suggestion of linking with `libflux.la`.  I see we were already doing that in the broker to access private message functions in libflux, FWIW.  I did notice that jobtap plugins don't explicitly link with `libflux-core.la` so we will have to add that I guess.

Reply:
Nothing too concerning IMHO on the executable size:
```
Before:

616K src/cmd/.libs/flux-exec
528K src/cmd/.libs/flux-start
1.7M src/modules/.libs/job-exec.so
1.3M src/modules/.libs/job-ingest.so
1.1M src/modules/.libs/sdbus.so

After:

616K src/cmd/.libs/flux-exec
528K src/cmd/.libs/flux-start
2.3M src/modules/.libs/job-exec.so
1.3M src/modules/.libs/job-ingest.so
1.8M src/modules/.libs/sdbus.so
```
I pushed this change as a separate commit - I'll squash it down it down if we agree this is OK.

Reply:
I find it _slightly_ preferable, but don't feel strongly enough to argue against the previous version if you prefer that. :shrug:

Reply:
This way is slightly preferable for me too, so I'll squash and set MWP.  Thanks!

Reply:
Nice work! :tada:

Reply:
Thanks!

I'm confused about something.  Consider the following snippet from `src/cmd/Makefile.am`

```
fluxcmd_ldadd = \
        $(top_builddir)/src/common/libkvs/libkvs.la \
        $(top_builddir)/src/common/librlist/librlist.la \
        $(top_builddir)/src/common/libflux-internal.la \
        $(top_builddir)/src/common/libflux-core.la \
        $(top_builddir)/src/common/libflux-optparse.la \
        $(FLUX_SECURITY_LIBS) \
        $(LIBPTHREAD) \
        $(JANSSON_LIBS)

flux_exec_LDADD = \
        $(top_builddir)/src/common/libsubprocess/libsubprocess.la \
        $(fluxcmd_ldadd) \
        $(top_builddir)/src/common/libflux/libflux.la
```
`flux-exec` needs the fbuf watcher in `libsubprocess.la`, which in turn needs `watcher.lo` functions like `watcher_create()`.   However, `watcher_create()` calls `flux_reactor_incref()`.

How does `flux_reactor_incref()` get resolved?  It could be resolved through `reactor.lo` in `libflux.la` or it could get resolved through the dynamic library `libflux-core.la`.  Since `libflux.la` is at the end of the list, I assumed it could not be resolved by something earlier in the list.  But somehow it works and the size of `flux-exec` is unchanged.

"I can't remember how the linker works, or if I ever knew how the linker works" seems to be a recurring theme for me :cry:

Reply:
I'd have to refresh my memory on this as well. I think the rule is something like "the linker reads libraries from left to right and accumulates unresolved symbols as it goes". This is why I suggested `libflux.la` should come last, since resolving symbols statically that are already in `libflux-core.so` should be the last resort.

But I also don't understand how the linker knows it needs `flux_reactor_incref()` in the example you pose above. Perhaps something else in the chain pulls that symbol in, so it is already there by the time `libflux.la` is processed? (I hope?) You could check the final `flux-exec` binary with `objdump` perhaps to see if it got pulled in statically.

Reply:
Oh duh, `libflux-core.so` is a dynamic library, so presumably all its symbols are available for anything before _or_ after on the link line. Maybe?

Reply:
That would imply that dynamic libs should always go first to maximize their use?  Hmm.  Thanks.  Signing off for tonight.